scrapy/core/scraper.py

```diff
@@ -142,7 +150,11 @@ def _scrape2(self, request_result, request, spider):
     def call_spider(self, result, request, spider):
         result.request = request
         dfd = defer_result(result)
-        dfd.addCallbacks(callback=request.callback or spider.parse,
+        callback = request.callback or spider.parse
+        if not six.PY2:
+            warn_on_generator_with_return_value(spider, callback, logger)
+            warn_on_generator_with_return_value(spider, request.errback, logger)
```

Comment:
Do you know if there is any pefrormance hit of doing code inspection on each request? I'm mostly concerned about pypy.

Reply:
Yup, I need to run some benchmarks. Stay tuned!

Reply:
Great idea. I often forget about `lru_cache`. The default maxsize is 128, which seems perhaps a bit too much (I don't think I have seen spiders with that many callbacks). I don't have a strong opinion on it, so feel free to suggest other values.

Reply:
Mind https://stackoverflow.com/questions/11815873/memoization-library-for-python-2-7

Reply:
Waiting until Python 3 makes more sense now :)

Reply:
Actually the whole PR works only in PY3, returns with values in generators are syntax errors in PY2

Reply:
That's why I'm skipping the tests in py2 and importing/defining the functions in py3 only.

Reply:
But since v1.7 was just released I can just remove all of those checks now :-)

Reply:
Right, I forgot about the `six.PY2` ifs!

> But since v1.7 was just released I can just remove all of those checks now :-)

Actually, there might be a 1.8, so don’t start removing `six` references just yet :slightly_smiling_face:

Reply:
Oh I just did hahaha. Nevermind, I can just revert the commit :-)